flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/SlotProfile.java

```diff
@@ -0,0 +1,275 @@
+		 */
+		@Nullable
+		<IN, OUT> OUT findMatchWithLocality(
+			@Nonnull Stream<IN> candidates,
+			@Nonnull Function<IN, SlotContext> contextExtractor,
+			@Nonnull Predicate<IN> additionalRequirementsFilter,
+			@Nonnull BiFunction<IN, Locality, OUT> resultProducer);
+	}
+
+	/**
+	 * This matcher implementation is the presence of prior allocations. Prior allocations are supposed to overrule
+	 * other locality requirements, such as preferred locations. Prior allocations also require strict matching and
+	 * this matcher returns null if it cannot find a candidate for the same prior allocation. The background is that
+	 * this will force the scheduler tor request a new slot that is guaranteed to be not the prior location of any
+	 * other subtask, so that subtasks do not steal another subtasks prior allocation in case that the own prior
+	 * allocation is no longer available (e.g. machine failure). This is important to enable local recovery for all
+	 * tasks that can still return to their prior allocation.
+	 */
+	@VisibleForTesting
+	public static class PreviousAllocationProfileToSlotContextMatcher implements ProfileToSlotContextMatcher {
+
+		/** Set of prior allocations. */
+		private final HashSet<AllocationID> priorAllocations;
+
+		@VisibleForTesting
+		PreviousAllocationProfileToSlotContextMatcher(Collection<AllocationID> priorAllocations) {
+			this.priorAllocations = new HashSet<>(priorAllocations);
+			Preconditions.checkState(
+				this.priorAllocations.size() > 0,
+				"This matcher should only be used if there are prior allocations!");
+		}
+
+		public <I, O> O findMatchWithLocality(
+			@Nonnull Stream<I> candidates,
+			@Nonnull Function<I, SlotContext> contextExtractor,
+			@Nonnull Predicate<I> additionalRequirementsFilter,
+			@Nonnull BiFunction<I, Locality, O> resultProducer) {
+
+			Predicate<I> filterByAllocation =
+				(candidate) -> priorAllocations.contains(contextExtractor.apply(candidate).getAllocationId());
+
+			return candidates
+				.filter(filterByAllocation.and(additionalRequirementsFilter))
+				.findFirst()
+				.map((result) -> resultProducer.apply(result, Locality.LOCAL)) // TODO introduce special locality?
+				.orElse(null);
+		}
+	}
+
+	/**
+	 * This matcher is used whenever no prior allocation was specified in the {@link SlotProfile}. This implementation
+	 * tries to achieve best possible locality if a preferred location is specified in the profile.
+	 */
+	@VisibleForTesting
+	public static class LocalityAwareRequirementsToSlotMatcher implements ProfileToSlotContextMatcher {
+
+		private final Collection<TaskManagerLocation> locationPreferences;
+
+		@VisibleForTesting
+		public LocalityAwareRequirementsToSlotMatcher(Collection<TaskManagerLocation> locationPreferences) {
+			this.locationPreferences = locationPreferences;
+		}
+
+		@Override
+		public <IN, OUT> OUT findMatchWithLocality(
+			@Nonnull Stream<IN> candidates,
+			@Nonnull Function<IN, SlotContext> contextExtractor,
```

Comment:
The extractor could return `TaskManagerLocation` to make it a bit more specific.

Reply:
I think we need to keep this so that we can obtain the allocation id in `PreviousAllocationProfileToSlotContextMatcher#findMatchWithLocality`.